### PR TITLE
DMP-1479 Global error handling

### PR DESCRIPTION
**Pull request by DavidTMann**
https://tools.hmcts.net/jira/browse/DMP-1479
https://tools.hmcts.net/jira/browse/DMP-1486

403 and 404 now handled globally. 

Updated subscribed endpoints array to include component handled responses. 

**Unhandled errors will default to 500 page.** 

- [X] commit messages are meaningful and follow good commit message guidelines
- [X] README and other documentation has been updated / added (if needed)
- [X] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/app/services/error/error-message.service.ts [hunk 1]
```diff
@@ -6,7 +6,11 @@ import { HeaderService } from '@services/header/header.service';
 import { BehaviorSubject, Observable } from 'rxjs';
 
 //Contains endpoints where errors are handled in their component
-const subscribedEndpoints = ['/api/cases/search'];
+const subscribedEndpoints = [
+  { endpoint: '/api/cases/search', responses: [204, 400, 500] },
+  { endpoint: 'api/audio-requests/playback', responses: [403] },
+  { endpoint: 'api/transcriptions', responses: [409] },
```
Comment (davet1985) [inline src/app/services/error/error-message.service.ts:12]: Maybe add a `/` at the start to be consistent?

File: src/app/services/error/error-message.service.ts [hunk 2]
```diff
@@ -43,14 +47,47 @@ export class ErrorMessageService {
...
+      const endpointMatches = response.url?.includes(subscribed.endpoint);
+      const responseMatches = subscribed.responses.includes(response.status);
+
+      return endpointMatches && responseMatches;
+    });
+  }
+
+  private showGlobalErrorPage(route: string) {
+    this.router.navigateByUrl(route);
     setTimeout(() => this.headerService.hideNavigation(), 0);
```
Comment (davet1985) [inline src/app/services/error/error-message.service.ts:79]: Is there a specific reason for the `setTimeout`?

File: src/app/services/error/error-message.service.spec.ts [hunk 3]
```diff
@@ -27,4 +27,31 @@ describe('ErrorMessageService', () => {
...
+  });
+
+  it('should route to internal-error page on unhandled error response, e.g. 409', () => {
+    const error = new HttpErrorResponse({ status: 409 });
+    const navigateSpy = jest.spyOn(mockRouter, 'navigateByUrl');
+
+    service.handleErrorMessage(error);
+
+    expect(navigateSpy).toHaveBeenCalledWith('internal-error');
+  });
```
Comment (davet1985) [inline src/app/services/error/error-message.service.spec.ts:56]: I might have missed it, but have you added a test for letting the error drop through? ie. you can check that `navigateSpy` has not been called when an endpoint/response matches the `subscribedEndpoints`?

**Review by davet1985 — COMMENTED** [review]
A few questions

File: src/app/services/error/error-message.service.ts (hunk 2 above)
Comment (DavidTMann) [reply src/app/services/error/error-message.service.ts:79]: Annoyingly it wouldn't work without setTimeout due to showNavigation being applied on URL changes by default (line 30 app.component.ts). So using setTimeout pushed it out to the next event loop to hideNavigation after.

File: src/app/services/error/error-message.service.spec.ts (hunk 3 above)
Comment (DavidTMann) [reply src/app/services/error/error-message.service.spec.ts:89]: A few were covered and updated in individual components, but have added all subscribed endpoints into error-message.service.spec now to check navigateSpy

File: src/app/components/search/search.component.spec.ts [hunk 4]
```diff
@@ -332,9 +336,11 @@ describe('SearchComponent', () => {
 
     flush();
 
+    // const navigateSpy = jest.spyOn(mockRouter, 'navigateByUrl');
```
Comment (davet1985) [inline src/app/components/search/search.component.spec.ts:339]: Remove comment

**Review by davet1985 — APPROVED** [review]

**Pull request merged**